fl4health/losses/deep_mmd_loss.py

```diff
@@ -0,0 +1,250 @@
+            hidden_size (int, optional): The hidden size of the deep network as the deep kernel used to compute
+                the MMD loss. Defaults to 10.
+            output_size (int, optional): The output size of the deep network as the deep kernel used to compute
+                the MMD loss. Defaults to 50.
+            lr (float, optional): Learning rate for training the Deep Kernel. Defaults to 0.001.
+            training (bool, optional): Whether the Deep Kernel is in training mode. Defaults to True.
+            is_unbiased (bool, optional): Whether to use the unbiased estimator for the MMD loss. Defaults to True.
+            gaussian_degree (int, optional): The degree of the generalized Gaussian kernel. Defaults to 1.
+            optimization_steps (int, optional): The number of optimization steps to train the Deep Kernel in each
+                forward pass. Defaults to 5.
+        """
+
+        super().__init__()
+        self.device = device
+        self.lr = lr
+        self.training = training
+        self.is_unbiased = is_unbiased
+        self.gaussian_degree = gaussian_degree  # generalized Gaussian (if L>1)
+        self.optimization_steps = optimization_steps
+
+        # Initialize the model
+        self.featurizer = ModelLatentF(input_size, hidden_size, output_size).to(self.device)
+
+        # Initialize parameters
+        self.epsilon_opt: torch.Tensor = torch.log(torch.from_numpy(np.random.rand(1) * 10 ** (-10)).to(self.device))
+        self.epsilon_opt.requires_grad = self.training
+        self.sigma_q_opt: torch.Tensor = torch.sqrt(torch.tensor(2 * 32 * 32, dtype=torch.float).to(self.device))
+        self.sigma_q_opt.requires_grad = self.training
+        self.sigma_phi_opt: torch.Tensor = torch.sqrt(torch.tensor(0.005, dtype=torch.float).to(self.device))
+        self.sigma_phi_opt.requires_grad = self.training
+
+        # Initialize optimizers
+        self.optimizer_F = torch.optim.AdamW(
+            list(self.featurizer.parameters()) + [self.epsilon_opt] + [self.sigma_q_opt] + [self.sigma_phi_opt],
+            lr=self.lr,
+        )
+
+    def pairwise_distiance_squared(self, X: torch.Tensor, Y: torch.Tensor) -> torch.Tensor:
+        """compute the paired distance between x and y."""
+        x_norm = (X**2).sum(1).view(-1, 1)
+        y_norm = (Y**2).sum(1).view(1, -1)
+        paired_distance = x_norm + y_norm - 2.0 * torch.mm(X, torch.transpose(Y, 0, 1))
+        paired_distance[paired_distance < 0] = 0
+        return paired_distance
+
+    def h1_mean_var_gram(
+        self,
+        k_x: torch.Tensor,
+        k_y: torch.Tensor,
+        k_xy: torch.Tensor,
+        is_var_computed: bool,
+    ) -> Tuple[torch.Tensor, Optional[torch.Tensor]]:
+        """compute value of MMD and std of MMD using kernel matrix."""
+        nx = k_x.shape[0]
+        ny = k_y.shape[0]
+
+        if self.is_unbiased:
+            # compute the unbiased MMD estimator (MMD_u^2) defined in Eq. (2) of the paper
```

Comment:
I think we want to be specific with what mmd is being calculated to match the paper, that is \hat{\text{MMD}}_u^2 (that is, which the hat).

Reply:
oh I see I updated to full notation now.